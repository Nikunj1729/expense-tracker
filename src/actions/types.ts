export const UPDATE_INCOME = "UPDATE_INCOME";
export const UPDATE_EXPENSE = "UPDATE_EXPENSE";
export const REMOVE_INCOME = "REMOVE_INCOME";
export const REMOVE_EXPENSE = "REMOVE_EXPENSE";
export const ADD_INCOME = "ADD_INCOME";
export const ADD_EXPENSE = "ADD_EXPENSE";
