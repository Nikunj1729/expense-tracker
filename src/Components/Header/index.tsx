import React from 'react';
import Box from '@material-ui/core/Box';
import logo from 'logo.png';

const Header = () => {
  return (
    <Box component="header" className="App-header sticky top-0 h-16 flex items-center px-6 bg-green-900">
      <img src={logo} className="App-logo h-8" alt="logo" />
    </Box>
  );
}

export default Header;
