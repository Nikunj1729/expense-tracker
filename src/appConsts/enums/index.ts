export enum Transations {
  INCOME = 'Income',
  EXPENSE = 'Expense',
  TRACKER = 'Tracker'
}