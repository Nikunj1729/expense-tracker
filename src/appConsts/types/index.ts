export type TransactionType = {
  name?: string;
  amount?: number;
  id?: string;
  type?: string;
};

export type ActionType = {
  type: string;
  payload: TransactionType;
};