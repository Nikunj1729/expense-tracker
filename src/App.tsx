import React, { useState, useEffect, Suspense } from 'react';
import { Provider } from 'react-redux';
import { I18nextProvider } from 'react-i18next';
import i18n from './i18n';
import Box from '@material-ui/core/Box';
import Snackbar from '@material-ui/core/Snackbar';
import CircularProgress from '@material-ui/core/CircularProgress';
import axios from 'axios';
import store from "./store";
import Header from './Components/Header';
import Chart from './Components/Chart';
import AddTransaction from './Components/AddTransactionForm';
import TransactionList from './Components/TransactionList';

const MyComponent = (props: any) => {
  const [loading, setLoading] = useState(true);
  const [error, setError] = useState(false);
  const [initState, setInitState] = useState({ transactions: {} });

  useEffect(() => {
    axios.get(`${process.env.REACT_APP_API_URL}/data`)
      .then(res => {
        setLoading(false);
        setInitState({ transactions: res.data });
      })
      .catch(err => {
        setLoading(false);
        setError(true)
      });
  }, []);

  if (loading) {
    return (
      <Box className="flex items-center justify-center w-full h-full">
        <CircularProgress />
      </Box>
    );
  }
  else if (error) {
    return <Snackbar
      anchorOrigin={{
        vertical: 'bottom',
        horizontal: 'left',
      }}
      open={true}
      autoHideDuration={6000}
      message="Error loading data from API. Please reload the application or try again."
    />;
  }
  else {
    return (
      <Provider store={store(initState)}>
        <Box className="App h-screen max-h-screen">
          <Header />
          <Box className="w-full h-full max-h-full-4 flex">
            <Chart />
            <Box className="w-6/12 h-full flex flex-col items-center justify-between py-16">
              <AddTransaction />
              <TransactionList />
            </Box>
          </Box>
        </Box>
      </Provider>
    );
  }

}

export default function App() {
  return (
    <Suspense fallback="loading">
      <I18nextProvider i18n={i18n}>
        <MyComponent />
      </I18nextProvider>
    </Suspense>
  );
}