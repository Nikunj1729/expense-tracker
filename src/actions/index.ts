import { ADD_INCOME, ADD_EXPENSE, UPDATE_INCOME, UPDATE_EXPENSE, REMOVE_INCOME, REMOVE_EXPENSE } from "./types";
import { TransactionType } from 'appConsts/types';

export const addIncome = (transaction: TransactionType) => {
  return {
    type: ADD_INCOME,
    payload: transaction
  };
};

export const addExpense = (transaction: TransactionType) => {
  return {
    type: ADD_EXPENSE,
    payload: transaction
  };
};

export const updateIncome = (transaction: TransactionType) => {
  return {
    type: UPDATE_INCOME,
    payload: transaction
  };
};

export const updateExpense = (transaction: TransactionType) => {
  return {
    type: UPDATE_EXPENSE,
    payload: transaction
  };
};

export const removeIncome = (id: string) => {
  return {
    type: REMOVE_INCOME,
    payload: { id }
  };
};

export const removeExpense = (id: string) => {
  return {
    type: REMOVE_EXPENSE,
    payload: { id }
  };
};
