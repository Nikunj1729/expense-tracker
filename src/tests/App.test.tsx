import React from 'react';
import * as actions from 'actions'
import * as types from 'actions/types'
import { shallow, mount, render } from 'enzyme';
import App from 'Components/Header';
import Header from 'Components/Header';
import reducer, { uid } from "reducers/transactionReducer";
import { Transations } from 'appConsts/enums';
import logo from 'logo.png';

describe('App', () => {

  it("renders without crashing", () => {
    shallow(<App />);
  });


  it('should render Header correctly with no props', () => {
    const component = shallow(<Header />);

    expect(component).toMatchSnapshot();
  });

  it("renders image header", () => {
    const header = render(<Header />);
    const image = <img alt="logo"
      className="App-logo h-8"
      src="logo.png" />;
    expect(header.find('img')).toHaveLength(1);
  })

  describe('actions', () => {
    it('should create an action to add a transaction', () => {
      const payload = { name: "Transaction", type: Transations.INCOME, amount: 5000 }
      const expectedAction = {
        type: types.ADD_INCOME,
        payload
      }
      expect(actions.addIncome(payload)).toEqual(expectedAction)
    })
  });

  describe('todos reducer', () => {
    it('should return the initial state', () => {
      expect(reducer(undefined, { type: '', payload: { name: "Transaction", type: Transations.INCOME, amount: 5000 } })).toEqual(
        {
          transactions: {}
        }
      )
    })
  });

  it('should handle ADD_INCOME', () => {
    expect(
      reducer({
        transactions: {}
      }, {
        type: types.ADD_INCOME,
        payload: { name: "Transaction", type: Transations.INCOME, amount: 5000 }
      })
    ).toMatchObject(
      {
        transactions: {
          [uid]: ["Transaction", "tracker", 5000]
        }
      }
    )
  })

  it('should handle ADD_INCOME', () => {
    expect(
      reducer({
        transactions: { [uid]: ["tracker", "Transaction", 5000] }
      }, {
        type: types.REMOVE_EXPENSE,
        payload: { id: uid }
      })
    ).toMatchObject(
      {
        transactions: {}
      }
    )
  })

});