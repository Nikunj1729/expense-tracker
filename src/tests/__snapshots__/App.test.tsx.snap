// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should render Header correctly with no props 1`] = `
<Styled(MuiBox)
  className="App-header sticky top-0 h-16 flex items-center px-6 bg-green-900"
  component="header"
>
  <img
    alt="logo"
    className="App-logo h-8"
    src="logo.png"
  />
</Styled(MuiBox)>
`;

exports[`App should render correctly with no props 1`] = `
<Styled(MuiBox)
  className="App-header sticky top-0 h-16 flex items-center px-6 bg-green-900"
  component="header"
>
  <img
    alt="logo"
    className="App-logo h-8"
    src="logo.png"
  />
</Styled(MuiBox)>
`;
